src/pybasic/tools/dct2d_tools.py

```diff
@@ -1,35 +1,99 @@
 
-import jax.numpy as jnp
+import os
+from abc import ABC, abstractmethod, abstractproperty
 
-from ..types import ArrayLike
+# TODO remove below after benchmarking or change to logging
+try:
+    import cv2
+except ImportError:
+    print("unable to import cv2")
+try:
+    import scipy.fft
+except ImportError:
+    print("unable to import scipy")
 
+import jax
+import numpy as np
 
-def dct2d(im_stack: ArrayLike) -> jnp.ndarray:
-    """Calculates 2D discrete cosine transform.
+__all__ = ["dct2d", "idct2d"]
+DEFAULT_BACKEND = "SCIPY"
 
-    Type -- DCT-II
 
-    Args:
-        im_stack: input image stack
+class DCT(ABC):
+    @abstractproperty
+    @abstractmethod
+    def _backend(self) -> str:
+        ...
 
-    Returns:
-        transformed image stack
-    """
-    ...
-    return
+    @staticmethod
+    @abstractmethod
+    def dct2d(arr: np.ndarray) -> np.ndarray:
+        ...
 
+    @staticmethod
+    @abstractmethod
+    def idct2d(arr: np.ndarray) -> np.ndarray:
+        ...
 
-def idct2d(im_stack: ArrayLike) -> jnp.ndarray:
-    """Calculates 2D inverse discrete cosine transform.
 
-    Type -- DCT-III
+class JaxDCT(DCT):
+    _backend = "JAX"
 
-    Args:
-        im_stack: input image stack
+    @staticmethod
+    def dct2d(arr: np.ndarray) -> np.ndarray:
+        return jax.scipy.fft.dct(jax.scipy.fft.dct(arr, norm="ortho"), norm="ortho")
 
-    Returns:
-        transformed image stack
-    """
-    ...
-    return
+    @staticmethod
+    def idct2d(arr: np.ndarray) -> np.ndarray:
+        return jax.scipy.fft.idct(
+            jax.scipy.fft.idct(arr, norm="ortho"), norm="ortho"
+        )  # FIXME only dct type 2 is implemented in JAX...
+
+
+class OpenCVDCT(DCT):
+    _backend = "OPENCV"
+
+    @staticmethod
+    def dct2d(arr: np.ndarray) -> np.ndarray:
+        return cv2.dct(arr)
+
+    @staticmethod
+    def idct2d(arr: np.ndarray) -> np.ndarray:
+        return cv2.dct(arr, flags=cv2.DCT_INVERSE)
+
+
+class SciPyDCT(DCT):
+    _backend = "SCIPY"
+
+    # NOTE can we use scipy.fftpack.dctn to replace the nested function below?
+
+    @staticmethod
+    def dct2d(arr: np.ndarray) -> np.ndarray:
+        return scipy.fft.dct(scipy.fft.dct(arr.T, norm="ortho").T, norm="ortho")
+
+    @staticmethod
+    def idct2d(arr: np.ndarray) -> np.ndarray:
+        return scipy.fft.idct(scipy.fft.idct(arr.T, norm="ortho").T, norm="ortho")
+
+
+DCT_BACKENDS = {"JAX": JaxDCT(), "OPENCV": OpenCVDCT(), "SCIPY": SciPyDCT()}
+
+
+if os.environ.get("BASIC_DCT_BACKEND"):
+    try:
+        dct = DCT_BACKENDS[os.environ["BASIC_DCT_BACKEND"]]
+    except KeyError as e:
+        # TODO change to logging or warning
+        raise e
+        # # OR when this fails, revert to default
+        # print(
+        #     f"unrecognized dct backend {os.environ['BASIC_DCT_BACKEND']}, "
+        #     f"defaulting to {DEFAULT_BACKEND}"
+        # )
+        # dct = DCT_BACKENDS[DEFAULT_BACKEND]
+else:
+    dct = DCT_BACKENDS[DEFAULT_BACKEND]

```

Comment:
I took this suggestion. It is tagged with a `TODO` to add logging.

Reply:
I need to troubleshoot asv because it is not working on my computer. I'll make an issue to write the benchmarks since this merged, but I'm going to try and get it working.